### ISSUE TITLE
Keep priority on completion?

### ISSUE THREAD

**Issue by lorentrogers** [samwho/todo-txt-gem#26]
Unless I'm not seeing it, I don't think [the spec][1] says that the priority needs to be removed once a task is complete? Is there a reason to remove it? Seems like it may be beneficial to keep it in the text portion of the task? (Do other apps generally remove it?)

[1]: https://github.com/ginatrapani/todo.txt-cli/wiki/The-Todo.txt-Format

**Comment by maetl** [comment]
The [completed task rules](https://github.com/ginatrapani/todo.txt-cli/wiki/The-Todo.txt-Format#rule-2-the-date-of-completion-appears-directly-after-the-x-separated-by-a-space) say the following:

> Many Todo.txt clients discard priority on task completion. To preserve it, use the key:value format described below (for example, pri:A)

The rationale for this is so completed tasks can be sorted by date rather than lexicographically:

> If you’ve prepended the creation date to your task, on completion it will appear directly after the completion date. This is so your completed tasks sort by date using standard sort tools.

**Comment by maetl** [comment]
So the implementation choices here are:

- Discard the priority (keep the current behaviour)
- Keep the priority by transforming it to `priority:A`/`pri:A` and injecting it back into the task line as a tag
- Providing a configuration option that flips between the two behaviours described above

**Comment by lorentrogers** [comment]
Sounds right to me - makes sense about why it's removed. 

Could we keep the priority regardless and just flip and option for how it's shown in `#to_s`?

**Comment by maetl** [comment]
Yeah, I don’t think there’s any reason to blow away the state. That way if we undo the task, it goes back to the way it was.

**Comment by lorentrogers** [comment]
Sounds good -- I can work on a PR for this. 

Until we have something like an `Options` class, should we go with `priority:A`, since it's more explicit than `pri:A`?

**Comment by lorentrogers** [comment]
Also, how should we handle this case?
```
(A) My top priority thing. pri:B priority:C
```
Would it override / strip out the existing fields? It's probably a rare case, but syntactically valid.

**Comment by lorentrogers** [comment]
I put together some basic logic for keeping the priority on completion in #33, but I'm not sure how to reconcile some of the other tests like this one: https://github.com/samwho/todo-txt-gem/blob/master/spec/todo-txt/task_spec.rb#L55

**Comment by lorentrogers** [comment]
Seems like the priority regex in `Syntax` should catch `(A)`, _and_ `priority:A`. But what do we do if both are present? Is there an expectation for how this is done in other apps?

**Comment by maetl** [comment]
That test looks incorrect to me, but I’ll have to check. I thought the general format guideline was that priorities get discarded on completion. In any case, the priority should be at the beginning of the line, not between the dates—so I’m not sure why that test is set up the way it is.

**Comment by lorentrogers** [comment]
Agreed that this is a strange test, and that it should probably be altered to match a more standard case. 

But, it does bring up a good issue -- if that string is syntactically correct, we should have a test to validate that it's parsed correctly. It wouldn't be a standard case, since the library we're working on removes the `(A)` formatted priority. But if it's valid, we should be sure that the library can still read the string and handle it well.

**Comment by lorentrogers** [comment]
Complete Tasks > Rule 2 does not specify that the priority must be removed, so I assume that it can stay, and would still be valid:
```
If you’ve prepended the creation date to your task, on completion it will appear directly after the completion date. This is so your completed tasks sort by date using standard sort tools. Many Todo.txt clients discard priority on task completion. To preserve it, use the key:value format described below (for example, pri:A)

With the completed date (required), if you’ve used the prepended date (optional), you can calculate how many days it took to complete a task.
```
